materialized: tls auth

benesch opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/materializeinc/materialize/5620)
<!-- Reviewable:end -->

benesch commented:
This is big, but 750 lines are just pretty verbose tests for all the various TLS modes. I'd be happy to try to split it up, but we'll just need to be careful if we do that to get it all in one release cycle, so that we don't end up with TLS auth that only authenticates, like, half the network protocols in a release.

benesch commented on doc/user/content/cli/_index.md:
```diff
@@ -150,28 +152,43 @@ the compaction window.
+`allow`       | Permits TLS encryption at the client's option.<br><br>Materialize serves both unencrypted and encrypted traffic over the same TCP port, as specified by [`--listen-addr`](#listen-address). The web UI will be served over HTTPS in addition to HTTP. Incoming SQL connections can negotiate TLS encryption; consult your SQL client's documentation for details.
+`require`     | Requires TLS encryption.<br><br>Materialize will reject HTTP connections and SQL connections that do not negotiate TLS.
+`verify-ca`   | Like `require`, but additionally requires that clients present a certificate.<br><br>Materialize verifies that the client certificate is issued by the certificate authority (CA) specified by the `--tls-ca` option.
+`verify-full` | Like `verify-ca`, but the Common Name (CN) field of the client certificate additionally determines the user who is connecting.<br><br>For HTTPS connections, this user is taken directly from the CN field. For SQL connections, the name of the user in the connection parameters must match the name of specified in the CN field.<br><br>This is the default mode if `--tls-cert` is specifi ...
```
Whoop, thanks, fixed.

benesch commented on doc/user/content/cli/_index.md:
````diff
@@ -150,28 +152,43 @@ the compaction window.
+options, respectively:
 
 ```shell
 $ materialized -w1 --tls-cert=server.crt --tls-key=server.key
````
Oh good point. Fixed.

benesch commented on doc/user/content/cli/_index.md:
```diff
@@ -186,12 +203,12 @@ OpenSSL 1.1.1g  21 Apr 2020
 
-Materialize configures OpenSSL according to Mozilla's [Modern
-compatibility][moz-modern] level, which requires TLS v1.3 and modern cipher
+Materialize configures OpenSSL according to Mozilla's [Intermediate
```
Oh it changed a while back for Fivetran support but I forgot to update the docs. Split out into #5685 with more details as requested below.

benesch commented on src/materialized/src/lib.rs:
```diff
@@ -176,9 +192,37 @@ pub async fn serve(
+        None => (None, None),
+        Some(tls_config) => {
+            let context = {
+                let mut builder = SslAcceptor::mozilla_intermediate_v5(SslMethod::tls())?;
```
Done in #5685.

benesch commented on src/materialized/tests/tls.rs:
```diff
@@ -0,0 +1,743 @@
+}
+
+fn run_tests<'a>(header: &str, server: &util::Server, tests: &[TestCase<'a>]) {
+    println!("==> {}", header);
```
Intentional actually; this is my hacky recreation of Go's subtests. It's friggin impossible to tell where one table-driven test case starts and ends otherwise. :/

benesch commented:
Thanks very much for the review! Addressed the first round of feedback.

petrosagg reviewed (approved):
Looks good to me!

The only thing that made me a bit uncomfortable was using `String` for the user type that could be potentially misused. What do you think of making a specific User/Username type that can only be constructed by the authentication logic plus some heavily documented and appropriate named methods to create a User out of thin air?

elindsey commented:
Change looks good to me by the way - just the 'allow' stuff we had chatted about previously. 🙂

benesch commented on src/materialized/src/bin/materialized/main.rs:
```diff
@@ -264,12 +321,29 @@ fn run(args: Args) -> Result<(), anyhow::Error> {
+            bail!("cannot specify --tls-mode=disable and --tls-cert simultaneously");
+        }
+        if args.tls_key.is_some() {
+            bail!("cannot specify --tls-mode=disable and --tls-key simultaneously");
```
Good point, done!

benesch commented:
Ok, addressed review feedback. `allow` is no longer a thing and the `--tls-ca` flag is properly rejected if used in `disable`  or `require`. 

@petrosagg I'm going to hold off on the `User` type for a moment, because making it _useful_ requires some changes to the HTTP auth flow. Those changes will be coming soon, though!

benesch commented:
Thanks for the reviews, all!
